thorbanks/utils.py

```diff
@@ -173,17 +180,42 @@ def verify_signature(request, bank_name, signature, auth=False, response=False):
 
     digest = request_digest(request, bank_name, auth=auth, response=response)
 
-    hash_algorithms = [hashes.SHA1(), hashes.SHA256()]
-    for hash_algorithm in hash_algorithms:
-        try:
-            public_key.verify(
-                b64decode(signature), digest, padding.PKCS1v15(), hash_algorithm
+    hash_algorithm = settings.get_verify_hash_algorithm(bank_name)
+    hasher = _get_hasher(hash_algorithm, bank_name)
+
+    try:
+        public_key.verify(b64decode(signature), digest, padding.PKCS1v15(), hasher)
+        return True
+    except InvalidSignature:
+        # Verification with the configured hash algorithm failed. We will try to verify
+        # with the other hash algorithms as a fallback in case the bank has switched
+        # to a new hash algorithm.
+        for hash_algorithm, hasher in HASH_ALGORITHMS.items():
+            try:
+                public_key.verify(
+                    b64decode(signature), digest, padding.PKCS1v15(), hasher
+                )
+                logging.warning(
+                    "%s banklink is using %s hash algorithm, "
+                    "but the banklink configuration is using %s.",
+                    bank_name,
+                    hash_algorithm,
+                    settings.get_hash_algorithm(bank_name),
+                )
+                return True
+            except InvalidSignature:
+                pass
+
+        if not auth:
+            # The client might have paid, but we don't accept the banklink response.
+            # Need to check if the payment was successful and contact the client!
+            logging.critical(
+                "Signature verification for the successful %s banklink callback has "
+                "failed. Immediate action required to confirm the payment's status.",
+                bank_name,
```

Comment:
This is potentially quite a serious issue but that was just logged as generic payment information before. I tried to make it a bit more clear what is happening. Now thinking about it, would be even better to bring this exception out to the app using this payments module and then it could be logged in Sentry as well with a clear error message to call people to action.

Reply:
this has been historically handled in the userland code (for example see gateme)

Reply:
but I guess having this exception here does not hurt. although I would look more of the payload contents here maybe (since unless you are using sentry you wont see the contents of the other vars)